### PR TITLE
Revise Branch Discovery Order for MSW Output

### PR DESCRIPTION
**Pull request by bska**
This commit places more conditions on the branch discovery order than the previous work in commit 6d3ee57dd (PR #3696).  In particular, we now ensure that branches are discovered ("created") in the order of increasing segment number of the branch outlet segment. Furthermore, if multiple branches have the same outlet segment, then we sort those branches on their increasing branch IDs.

To this end, switch from using a `std::queue<int>` to using a `std::priority_queue<KickOffPoint>`, with the `KickOffPoint` being a custom structure holding the kick-off segment, the branch outlet segment, and the branch ID and a custom `operator<()` to plug into the `priority_queue<>` heap mechanism.

This new sort order changes the result of certain unit tests, but those changes are expected and desired.

### CODE REVIEW

**Comment by bska** [comment]
I want to add some more unit testing here so this PR isn't quite ready for review/merging yet.  I'm creating the PR in draft mode to nevertheless enable testing.

**Comment by bska** [comment]
jenkins build this please

**Comment by tskille** [comment]
This PR has been tested on a full field model in Equinor are results are good. Looking forward to se this going into master

**Comment by bska** [comment]
> This PR has been tested on a full field model in Equinor are results are good. Looking forward to see this going into master

Thank you for testing.  I have amended the set of unit tests to cover some situations that we did not properly handle earlier and I'm marking this as "ready for review".

**Comment by bska** [comment]
jenkins build this please

File: src/opm/output/eclipse/AggregateMSWData.cpp [hunk 1]
```diff
@@ -1007,6 +1007,24 @@ namespace {
...
+                bool operator<(const KickOffPoint& that) const
+                {
+                    // We use '>' here to have low outlet segment IDs move
+                    // to the front of the priority queue.  If multiple
+                    // branches kick off from the same outlet segment, we
+                    // sort these by their branch IDs whence lower-ID
+                    // branches are created before higher-ID branches.
+                    return std::tie(this->outlet, this->branch)
+                        >  std::tie(that .outlet, that .branch);
+                }
```
Comment (blattms) [inline src/opm/output/eclipse/AggregateMSWData.cpp:1025]: This does not seem very expressive. If I am not mistaking then this is an operator< that actually performs the operation of operator>. When this is used people will be quite surprised.

Wouldn't is be better to call it `operator>` and e.g. use `std::greater<KickOffPoint>` as an additional template parameter for `std::priority_queue`?

**Review by blattms — COMMENTED** [review]
Looks good. Only KickOffPoint::operator< seems a bit surprising.

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 1 above)
Comment (blattms) [reply src/opm/output/eclipse/AggregateMSWData.cpp:1025]: I guess this is not worth the time to discuss further. It is  a private class.

**Comment by blattms** [comment]
jenkins build this please

File: src/opm/output/eclipse/AggregateMSWData.cpp (hunk 1 above)
Comment (bska) [reply src/opm/output/eclipse/AggregateMSWData.cpp]: > I guess this is not worth the time to discuss further. It is a private class.

Sure, but I appreciate the comment.  I've updated the PR accordingly.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
PR approved and build check is green.  I'll merge into master.  Thanks a lot, @blattms, for your suggestion&ndash;it absolutely improved this work.

**Pull request merged**